src/libmongoc/tests/test-mongoc-write-concern.c

```diff
@@ -595,10 +596,150 @@ test_write_concern_wtimeout_preserved (void)
    ASSERT (bson_iter_next (&child));
    ASSERT_CMPSTR (bson_iter_key (&child), "wtimeout");
    ASSERT_CMPINT64 (bson_iter_int64 (&child), ==, INT64_MAX);
-   
+
    mongoc_write_concern_destroy (write_concern);
 }
 
+/* callback that records write concern for commands */
+static void
+write_concern_count (const mongoc_apm_command_started_t *event)
+{
+   const bson_t *command = mongoc_apm_command_started_get_command (event);
+   bson_iter_t iter, iter_rec;
+
+   if (bson_iter_init (&iter, command)) {
+      int *sent_collection_w =
+         (int *) mongoc_apm_command_started_get_context (event);
+      if (bson_iter_find_descendant (&iter, "writeConcern.w", &iter_rec)) {
+         *sent_collection_w = bson_iter_int32 (&iter_rec);
+      } else {
+         *sent_collection_w = 1; /* no write concern sent. default used */
+      }
+   }
+}
+
+/* tests that unacknowledged write concerns are inherited from the
+ * collection in fam with options.  Addresses concerns from CDRIVER-3595 */
+void
+test_write_concern_inheritance_fam (void *context)
+{
+   mongoc_find_and_modify_opts_t *opts;
+   bson_t *update;
+   bson_error_t error;
+   bson_t query = BSON_INITIALIZER;
+   mongoc_write_concern_t *wc;
+   bool success;
+   int sent_w = -1;
+   mongoc_apm_callbacks_t *callbacks;
+   mongoc_client_t *client = test_framework_client_new ();
+   mongoc_collection_t *collection =
+      mongoc_client_get_collection (client, "db", "collection");
+
+   BSON_APPEND_UTF8 (&query, "test", "test");
+   update = bson_new ();
+
+   callbacks = mongoc_apm_callbacks_new ();
+   mongoc_apm_set_command_started_cb (callbacks, write_concern_count);
+   mongoc_client_set_apm_callbacks (client, callbacks, (void *) &sent_w);
+
+   opts = mongoc_find_and_modify_opts_new ();
+   mongoc_find_and_modify_opts_set_update (opts, update);
+
+   wc = mongoc_write_concern_new ();
+
+   mongoc_write_concern_set_w (wc, MONGOC_WRITE_CONCERN_W_UNACKNOWLEDGED);
+   mongoc_collection_set_write_concern (collection, wc);
+
+   success = mongoc_collection_find_and_modify_with_opts (
+      collection, &query, opts, NULL, &error);
+
+   BSON_ASSERT (success);
+   BSON_ASSERT (sent_w == MONGOC_WRITE_CONCERN_W_UNACKNOWLEDGED);
+
+   mongoc_write_concern_destroy (wc);
+   mongoc_apm_callbacks_destroy (callbacks);
+   mongoc_collection_destroy (collection);
+   mongoc_client_destroy (client);
+   bson_destroy (update);
+   bson_destroy (&query);
+   mongoc_find_and_modify_opts_destroy (opts);
+}
+
+/* tests that in fam operations, with a transaction in progress,
+ * write concerns are not inherited from collections.  Adresses concerns
+ * in CDRIVER-3595 */
+void
+test_write_concern_inheritance_fam_txn (void *context)
```

Comment:
If this test is structurally similar to the other, I suggest consolidating them and parameterizing on a `bool in_txn`.